sklearn/preprocessing/_encoders.py

```diff
@@ -764,10 +776,51 @@ def fit(self, X, y=None):
         self
 
         """
-        # base classes uses _categories to deal with deprecations in
-        # OneHoteEncoder: can be removed once deprecations are removed
-        self._categories = self.categories
-        self._fit(X)
+        X_temp = check_array(X, force_all_finite=False, dtype=None)
+        if not hasattr(X, 'dtype') and np.issubdtype(X_temp.dtype, np.str_):
+            X = check_array(X, force_all_finite=False,
+                            dtype=np.object)
+        else:
+            X = X_temp
+
+        n_samples, n_features = X.shape
+
+        if self.categories != 'auto':
+            if X.dtype != object:
+                for cats in self.categories:
+                    if not np.all(np.sort(cats) == np.array(cats)):
+                        raise ValueError("Unsorted categories are not "
+                                         "supported for numerical categories")
+            if len(self.categories) != n_features:
+                raise ValueError("Shape mismatch: if n_values is an array,"
+                                 " it has to be of shape (n_features,).")
+
+        self.categories_ = []
+
+        for i in range(n_features):
+            Xi = X[:, i]
+            if self.categories == 'auto':
+                try:
+                    cats = _encode(Xi)
+                except TypeError:
+                    cats = _encode(Xi.astype('<U3'))
+            else:
+                cats = np.array(self.categories[i], dtype=X.dtype)
+                diff = _encode_check_unknown(Xi, cats)
+                if diff:
+                    msg = ("Found unknown categories {0} in column {1}"
+                           " during fit".format(diff, i))
+                    raise ValueError(msg)
+            self.categories_.append(cats)
+
+        # Removing missing values from each column
+        for i, cats in enumerate(self.categories_):
+            try:  # find missing for numeric dtype
+                missing_i = np.isnan(cats.astype(float), casting='unsafe')
+                self.categories_[i] = cats[~missing_i]
+            except (ValueError, TypeError):  # find missing for object dtype
+                missing_i = (cats == 'nan')
```

Comment:
no () needed

Reply:
I don't think we should be treating the string 'nan' as signifying np.nan... we might need to do this checking before or in `_encode`